### ISSUE TITLE
Batch execution mode (non-REPL runs)

### ISSUE THREAD

**Issue by tizoc** [Shen-Language/shen-sources#4]
# Batch execution mode discussion

At the moment Shen assumes that the user program is operating from a REPL. But sometimes programs have to run in batch mode (like when invoked from the command line without being loaded from the REPL for example, as in `shen somescript.shen`.

The proposal is to add a new entry point (named `shen.run-batch` or similar) that changes some behaviors (like how to handle partial functions, in the REPL the user is asked if he wants to track the function, in batch mode this should just raise an error and quit).
### Handling of partial functions when there is no case to handle the passed arguments

There are at least two options, one is to compile code differently when in batch execution mode, the other is to have the default case handler behave differently at runtime depending on which mode Shen is running.

**Quote from an email in the original thread [1]**

> When compiling define's cases Shen generates a default case in case that will be called if the cases defined by the user are not exhaustive and the function is called with a value that is not matched by them.
> 
> https://github.com/Shen-Language/shen-sources/blob/17fb1103cfd290ba741de57d8524edec06a7f8c8/sources/track.shen#L33
> 
> default case generated here (and functions defined before in that file)
> 
> https://github.com/Shen-Language/shen-sources/blob/eb64276a74cca00cbe991fa1e6d76ea2caf8ba63/sources/core.shen#L391
> 
> One solution is to have a global variable switch so that when in batch mode shen generates a different default case. Another (easier) solution is to have a global variable switch that f_error will check for before asking for before printing the tracking prompt (and instead just raising an error).
#### Option 1. Different compilation when in batch execution mode

This is a static solution, the change will happen at compile time.

**Pros**: Can't think if any right now.

**Cons**: 
- Code that has been compiler for batch runs will not behave as expected when laded in the REPL (in implementations that support loading precompiled code) and vice versa.
#### Option 2. Have the default case handler function decide at runtime what to do

This is a dynamic solution. A global variable containing the execution mode (with one of `interactive` and `batch` as a vaule for example) which will be consulted by the default handler function to decide what to do.

**Pros**:
- In implementations that support loading precompiled code, it will always behave as expected.

**Cons**: Can't think if any right now.
#### Option 3. Store the default case handler in a variable

This is the most dynamic option, a global variable would contain a pointer to a lambda to be called to handle the default case in partial functions

**Pros**:
- More flexible, lets the user use custom handlers.

**Cons**:
- Could be a bit slower (but speed is not an issue here)
### New entry point that doesn't launch the REPL

Shen's entry point is the function `shen.shen`, it launches the REPL.

For batch mode a new entry point has to be defined.

_TODO: elaborate this more_
## References
- [1] https://groups.google.com/d/msg/qilang/4AxtzPlwi74/RVFoD_1pAgAJ

**Comment by gravicappa** [comment]
Want to comment about syntax:
```
Support for batch processing: shen.exe: [-t] [-e string] [-l filename]

-t : toggle type checking
-e : eval, print, exit
-l : load script, exit
```
I'd suggest the form `shen [-noinit] [-t] [-e string] [filename ...]`.
Taking filename(s) without additional keys allows making shebang Shen-scripts:
```
#!/usr/bin/env shen
(output "Ku!~%")
```
[Shen run](https://github.com/gravicappa/shen_run) employs this feature allowing be to bootstrap shen-js from [make.shen](https://github.com/gravicappa/shen-js/blob/master/make.shen).

**Comment by gravicappa** [comment]
Also it is essential to have `stderr`-like output stream with all errors written to it instead of `(stoutput)`.
I don't think that running `./some-shen-program >data` and having some `something something error` line in `data` is a good thing.

**Comment by tizoc** [comment]
@gravicappa agree.

I would treat those two in separate issues though, to keep the scope of each narrow.

`stderr` was discussed on the list already long ago, I have to find that thread to use for reference. It should be a quickly resolved issue, we probably don't need anything more than a new global variable and a function. There were other things that we discussed that should become issues here (I remember other I/O related discussions, like being able to set positions in streams, open files in append or truncate modes, etc)

Regarding the command line flags, I don't know if that is something that needs to be in the kernel, as long as the kernel provides enough means for it to be implemented (this, and #3 are intended to solve that) then the the `shen_run` functionality could be a library (or a command like now that is port agnostic, but takes advantage of the new entry point we would be adding here plus what is discussed in #3 instead of I/O redirection). That way programs that are compiled to be executed in a standalone way (without having to do `shen program.shen`) don't have to worry about including this stuff.

Thoughts on that?

**Comment by NHALX** [comment]
Shen could just provide a new file "interface.shen", with some toplevel primitives for backend authors to generate whatever type of entry point they need. 

untested pseudocode:
```
(define shen.interactive
  - -> undefined \\ enable hush, partial function errors, etc
  + -> undefined)

\\ simple batch interface:
(define shen.batch-run
  Action Handler -> (let Saved (interactive -)
                      (trap-error
                       (let Result (Action Saved)
                         (do 
                          (interactive Saved)
                          Result))
                       
                       (/. E (do
                              (interactive Saved)
                              (Handler E))))))



\\ Toplevel actions

(define shen.read-eval-print-loop
  TC -> (do (set *tc* TC) (shen.shen)))

(define shen.file-eval
  TC X -> (do (set *tc* TC) (load X)))

(define shen.string-eval-print
  TC X -> (let
              Read (read-from-string X)
              Eval (toplevel_evaluate Read TC)
              Show (make-string "~S~%" Eval)
            (pr Show)))

\** test **\

(shen.batch-run 

 (lambda _ 
   (do
    (shen.string-eval-print true "(+ 1 2)")
    (shen.file-eval false "test.shen")))

 (lambda E 
   (pr (error-to-string E) (stoutput))))
```

**Comment by NHALX** [comment]
Also note that the working dir should probably be changed before the load in (file-eval)

**Comment by deech** [comment]
What would be the downside of delegating this to the host platform? For example, leave the kernel the way it is and use the host language to call the "main" Shen function. That way the core can stay small.

**Comment by tizoc** [comment]
The way I was thinking about it, ports would handle it, the only thing that would change in the kernel is the ability to switch modes (either with something similar to how typechecking is enabled/disabled, or by providing a new entry point).

This is not something the ports can do 100% by themselves because some internals have to change a bit in the kernel (the tracking prompt for example, not sure what else yet).

@deech just to clarify (because this was not discussed anywhere yet) for the issues I opened, assume by default that my intention is to make the smallest possible change to the kernel (if at all), because 1) I want to keep it small too, 2) we probably don't have enough time and energy to make and test big changes (at least, I know that I myself don't for now).

I'm only interested in bug fixes, small changes and in agreeing in common interfaces between port maintainers, everything else should probably be an external library (with maybe a few small updates to the kernel if required to make it more extendable "from the outside").

**Comment by rkoeninger** [comment]
I have both the [shen-cl](https://github.com/Shen-Language/shen-cl) and [ShenSharp](https://github.com/rkoeninger/ShenSharp) ports handling command line arguments like `-l` for preloading files and `-e` for evaluating expressions, allowing one to do this: `shen -l make.shen -e "(make)"`.

In the course of enabling this, I noticed how the code is repeated between the two: [primitives.lsp](https://github.com/Shen-Language/shen-cl/blob/master/primitives.lsp#L291-L324), [RunRepl.fs](https://github.com/rkoeninger/ShenSharp/blob/master/Shen.Repl/RunRepl.fs#L19-L47).

Since the `*argv*` symbol and `command-line` function are now available in the kernel, the port could just set the args to `*argv*` and let the `shen.shen` function decide whether to start the REPL or what else to do.

This way, we'd have consistent command-line behavior across ports, and it would be available by the port adding 1 line of code.

**Comment by deech** [comment]
I think consistent experience at this level across ports is tempting but a siren song. I envision each port leveraging the strengths of its host, in this case the host's CLI functions, without having to worry about compatibility. I'm personally in favor a kernel that only implements Shen's core semantics and leaves all IO and other concerns to the host or to libraries. I would, for example, be in favor of removing the `read-byte` and `write-byte` primitives and all the kernel functions that depend on them because, in my opinion, they aren't well thought out.

**Comment by tizoc** [comment]
> I would, for example, be in favor of removing the read-byte and write-byte primitives and all the kernel functions that depend on them because, in my opinion, they aren't well thought out.

@deech can you elaborate on this? I also think that I/O is one of Shen's big weak spots, I just can't think how the kernel could have done better without overcomplicating things too much (either by requiring too much from ports, or by not doing anything at all.... again requiring too much from ports). Maybe one day the REPL can be split out from the kernel, and in that case I/O will not be required by the kernel to work, but that doesn't seem too likely in the near future.

As for the other stuff, I agree with the general sentiment. Anything that can be a library, is better done as a library, with changes in kernel being for performance, new hooks that open new possibilities, bug fixes, etc.

For I/O, networking and subprocesses, my intention is to just implement something that works well with the Scheme port and then publish it, others can then look at it and see if the API is good enough for other ports or not, and if not, what can be done better, eventually we may reach a common API. All that can be done without having to add such stuff to the kernel.

I know the libraries experience right now isn't good (or inexistent), so pushing stuff in the kernel is feels more convenient, but doing so will just make things harder in the future.

**Comment by tizoc** [comment]
This is solved by #81, maybe not 100% (handling of partial functions for example), but mostly.

For what remains I will open another issue that has more about better delimitation between interactive and static features of the kernel (part of that has been solved by the "dynamic-expand" extension, but there is more that can be done).
